### ISSUE TITLE
Error when adding multiple modules that share some source code

### ISSUE THREAD

**Issue by ElChache** [cursive-ide/cursive#2078]
When trying to add a module that shares some source code with other module, it shows an error message _Content root "/opt/libs/shared-lib/src" is defined for modules "module1" and "module2".Two modules in a project cannot share the same content root_.

It happens when two projects share the same library, by setting : {:local/root "../shared-lib" :deps/manifest :deps} on :deps in the deps.edn file

The workaround for now is to remove the code from the _Root content_  section on Project Structure -> Project Settings -> Modules, and click on apply.

But everytime a new module is added, even if the new module doesn't have shared modules, it complains again.

This makes it very tedious to manage multiple modules in the same project

**Comment by cursive-ide** [comment]
I can't reproduce this - when I try this, Cursive creates one library which is shared by the two modules.

Based on that message, it looks like you're adding the `src` folders from the shared library manually to module1 and module2, is that correct? If so, that's not required - the library sources will be automatically created and linked by Cursive.

**Comment by ElChache** [comment]
On this scenario module1 and module2 are not related, and there is no dependencies between them. They just happen to share a library. I could always open module1 in one project, and module2 in another, but I find very convenient to work with multiple modules on the same project.

So yes, in this scenario I have to add module2 manually, as there are no dependencies on module1 that would add module2 automatically.

When I add module2, it will complain with the message _Content root "/opt/libs/shared-lib/src" is defined for modules "module1" and "module2"_. Then, if I remove the shared library from Content Root, it doesn't complain anymore, and lets me apply. Then it seems like the library is again added to  Content Root silently, and I can work with both modules on the same project.

But as I keep adding modules (again, manually), I have to go through all the existent modules and remove the library from Content Root.

So, it seems like the only issue is the message on the Project Settings -> Modules. If you patiently remove all the shared code from Content Root, once you apply, the shared code seems to be  added automatically again. But you can imagine that if you have multiple modules sharing a bunch of local libraries, removing all the shared code on each module can be quite a lot of manual work.

**Comment by arichiardi** [comment]
Throwing in here, can this be related? https://dev.clojure.org/jira/browse/TDEPS-74

**Comment by cursive-ide** [comment]
@ElChache Right, I understand. Could you add a screenshot of the error you're seeing and your Project Structure window afterwards showing the shared lib you have to remove from Content Roots? I'd like to make sure it's matching what I'm seeing here.

**Comment by ElChache** [comment]
Step 1 - Just added a module that uses some libraries that will be shared by other modules. Shared modules don't show now, I'd have to re-open this window to see them.
![cursive issue step 1](https://user-images.githubusercontent.com/2159114/48875454-fd52fe80-edad-11e8-956d-30d5c276955c.png)

Step 2 - Added a second module. The shared libraries don't show now neither (but they show now for the previous module), and clicking in apply succeeds
![cursive issue step 2](https://user-images.githubusercontent.com/2159114/48875466-1956a000-edae-11e8-86bb-e4a7cc61f69d.png)

Step 3 - Added a third module. Clicking in apply now will fail because I already have 2 previous modules with shared libraries.
![cursive issue step 3](https://user-images.githubusercontent.com/2159114/48875572-9124ca80-edae-11e8-844a-bf6ecf30ba45.png)

Step 4 - Error message. Every time a new module is inserted from now on, it will show error messages for every library shared by any two modules
![cursive issue step 4](https://user-images.githubusercontent.com/2159114/48875623-dd700a80-edae-11e8-8c2a-73bbb134d354.png)

The workaround now is to go through each module and remove the shared libraries from Content Root. They are silently added later on.

**Comment by cursive-ide** [comment]
So in this case, somehow `dev` is being expanded to `/home/davidcerezoinigo/dev` but that is incorrect, right? I assume that should be `<project_home>/dev`? Which version of Cursive are you using?

**Comment by ElChache** [comment]
right right! I am using 1.8.0-2018.1

**Comment by cursive-ide** [comment]
So I've spent some time looking into this, and I cannot figure it out. Because of the details of how this is implemented, it's hard to get logging output. I've asked JetBrains for help, and I'll let you know when I have a better idea of how to investigate further.

**Comment by nblumoe** [comment]
I am having this issue too. Here is a minimal repo to reproduce the error: https://github.com/nblumoe/cursive-source-root-conflicts

Instructions:
Import projects recursively
![Capto_Capture 2019-06-25_09-13-50_AM](https://user-images.githubusercontent.com/995628/60078021-6d05df00-972b-11e9-8d53-1bae78e7e01d.png)

Check module settings, both `app` modules will have an identical source path to the `shared` module's src:
![Capto_Capture 2019-06-25_09-18-35_AM](https://user-images.githubusercontent.com/995628/60078079-8b6bda80-972b-11e9-9600-90676a03b802.png)
![Capto_Capture 2019-06-25_09-18-51_AM](https://user-images.githubusercontent.com/995628/60078094-8eff6180-972b-11e9-8cd1-d8e02cf34353.png)

The module config cannot be modified without deleting the shared path first (on rescanning the project files, Cursive will add the shared folder again):
![Capto_Capture 2019-06-25_09-18-04_AM](https://user-images.githubusercontent.com/995628/60078102-932b7f00-972b-11e9-9f25-57634ce1dadd.png)

**Comment by nblumoe** [comment]
@cursive-ide is this something you are going to look into?

**Comment by cursive-ide** [comment]
@nblumoe Sorry for the delay, yes, I'll try to look at this shortly.

**Comment by cursive-ide** [comment]
@nblumoe My apologies for taking so long to look at this. Your problem is different to the original one in this issue - that one is using Deps, and seemed to be due to Cursive calculating the shared paths incorrectly.

Your case is caused by the two modules explicitly trying to share the same source root. IntelliJ disallows this, as the message explains. There's really no good way around this restriction.

Normally what I would suggest is that you make the shared module a dependency of the two app modules, using: 
```
:dependencies [[org.clojure/clojure "1.10.0"]
               [shared "0.1.0-SNAPSHOT"]]
```
Cursive is smart enough to detect this case, and will create a module dependency between the app modules and the shared one, so you would be able to access the source code directly. However in your case, you're referencing the source as part of your `cljsbuild` config so that won't work. I don't really have a good solution to suggest for that case, unfortunately. The only thing I can think of would be to symlink the shared code from the two app modules - ugly, but it might work. Sorry I don't have a better answer for you.
